docarray/utils/lookup.py

```diff
@@ -0,0 +1,325 @@
+            f' is not supported'
+        )
+
+
+## Classes to compose compound lookups (Q object)
+
+
+class LookupTreeElem(object):
+    """Base class for a child in the lookup expression tree"""
+
+    def __init__(self):
+        self.negate = False
+
+    def evaluate(self, item):
+        raise NotImplementedError
+
+    def __or__(self, other):
+        node = LookupNode()
+        node.op = 'or'
+        node.add_child(self)
+        node.add_child(other)
+        return node
+
+    def __and__(self, other):
+        node = LookupNode()
+        node.add_child(self)
+        node.add_child(other)
+        return node
+
+
+class LookupNode(LookupTreeElem):
+    """A node (element having children) in the lookup expression tree
+
+    Typically it's any object composed of two ``Q`` objects eg::
+
+        >>> Q(language__neq='Ruby') | Q(framework__startswith='S')
+        >>> ~Q(language__exact='PHP')
+
+    """
+
+    def __init__(self, op: Union[str, bool] = 'and', negate: bool = False):
+        super(LookupNode, self).__init__()
+        self.children: List[LookupNode] = []
+        self.op = op
+        self.negate = negate
+
+    def add_child(self, child):
+        self.children.append(child)
+
+    def evaluate(self, doc: 'BaseDocument'):
+        """Evaluates the expression represented by the object for the document
+
+        :param doc : the document to match
+        :return: returns true if lookup passed
+        """
+        results = map(lambda x: x.evaluate(doc), self.children)
+        result = any(results) if self.op == 'or' else all(results)
+        return not result if self.negate else result
+
+    def __invert__(self):
+        newnode = LookupNode()
+        for c in self.children:
+            newnode.add_child(c)
+        newnode.negate = not self.negate
+        return newnode
+
+    def __repr__(self):
+        return f'{self.op}: [{self.children}]'
+
+
+class LookupLeaf(LookupTreeElem):
+    """Class for a leaf in the lookup expression tree"""
+
+    def __init__(self, **kwargs):
+        super(LookupLeaf, self).__init__()
+        self.lookups = kwargs
+
+    def evaluate(self, doc: 'BaseDocument'):
+        """Evaluates the expression represented by the object for the document
+
+        :param doc : the document to match
+        :return: returns true if lookup passed
+        """
+        result = all(lookup(k, v, doc) for k, v in self.lookups.items())
+        return not result if self.negate else result
+
+    def __invert__(self):
+        newleaf = LookupLeaf(**self.lookups)
+        newleaf.negate = not self.negate
+        return newleaf
+
+    def __repr__(self):
+        return f'{self.lookups}'
+
+
+# alias LookupLeaf to Q
+Q = LookupLeaf
+
+
+## Exceptions
+
+
+class LookupyError(Exception):
+    """Base exception class for all exceptions raised by lookupy"""
+
+    pass
+
+
+## utility functions
+
+
+def dunder_partition(key):
```

Comment:
these utility functions probably come from the other library, but should we add type hints to them? we are pretty strict about that in the rest of the codebase

Reply:
I will try to do so